--pretty output is confusing

<!-- BUGS: Please use this template. -->
<!-- QUESTIONS: This is not a general support forum! Ask Qs at http://stackoverflow.com/questions/tagged/typescript -->
<!-- SUGGESTIONS: See https://github.com/Microsoft/TypeScript-wiki/blob/master/Writing-Good-Design-Proposals.md -->

**TypeScript Version:**  2.3.4

When the compiler outputs errors it uses this format:
```
EXCERPT 1
   ^^   1

FILE: ERROR 1
EXCERPT 2
  ^^    2

FILE: ERROR 2
```

However, this is confusing since the first error is more closely associated with the second excerpt since it is separated by an empty line and a mostly blank line. I would propose changing it to:

```
FILE: ERROR 1
EXCERPT 1
   ^^   1

FILE: ERROR 2
EXCERPT 2
  ^^    2

```

I find the ordering here, with excerpt above error, really unintuitive. Every other compiler I can think of to check does it in the other order:

javac
```
Test.java:3: error: cannot find symbol
  public static void main(Stirng args[]) {
                          ^
```

clang
```
test.cc:4:3: error: use of undeclared identifier 'cuot'
  cuot << "hello";
  ^
```

gcc
```
test.cc: In function ‘int main()’:
test.cc:4:3: error: ‘cuot’ was not declared in this scope
   cuot << "hello";
   ^
```

Closure compiler:
```
input0:1: WARNING - Property lgo never defined on Console
console.lgo('hello');
        ^^^
```

Swift:
```
test.swift:1:1: error: use of unresolved identifier 'prnit'
prnit("Hello, world!")
^~~~~
```

> For some context, the original design was to mimick how tooltips appear below the error span.

But...they don't, at least not in VS Code, unless you're on the very first line.

![image](https://user-images.githubusercontent.com/782920/27973896-abc1de0c-6329-11e7-8291-e0420b6408eb.png)

--- a/src/compiler/program.ts
+++ b/src/compiler/program.ts
@@ -271,6 +271,7 @@ namespace ts {
     export function formatDiagnosticsWithColorAndContext(diagnostics: ReadonlyArray<Diagnostic>, host: FormatDiagnosticsHost): string {
         let output = "";
         for (const diagnostic of diagnostics) {
+            let context = "";
             if (diagnostic.file) {
                 const { start, length, file } = diagnostic;
                 const { line: firstLine, character: firstLineChar } = getLineAndCharacterOfPosition(file, start);
@@ -284,12 +285,12 @@ namespace ts {
                     gutterWidth = Math.max(ellipsis.length, gutterWidth);
                 }
 
-                output += host.getNewLine();
+                context += host.getNewLine();
                 for (let i = firstLine; i <= lastLine; i++) {
                     // If the error spans over 5 lines, we'll only show the first 2 and last 2 lines,
                     // so we'll skip ahead to the second-to-last line.
                     if (hasMoreThanFiveLines && firstLine + 1 < i && i < lastLine - 1) {
-                        output += formatAndReset(padLeft(ellipsis, gutterWidth), gutterStyleSequence) + gutterSeparator + host.getNewLine();
+                        context += formatAndReset(padLeft(ellipsis, gutterWidth), gutterStyleSequence) + gutterSeparator + host.getNewLine();
                         i = lastLine - 1;
                     }
 
@@ -300,30 +301,28 @@ namespace ts {
                     lineContent = lineContent.replace("\t", " ");    // convert tabs to single spaces
 
                     // Output the gutter and the actual contents of the line.
-                    output += formatAndReset(padLeft(i + 1 + "", gutterWidth), gutterStyleSequence) + gutterSeparator;
-                    output += lineContent + host.getNewLine();
+                    context += formatAndReset(padLeft(i + 1 + "", gutterWidth), gutterStyleSequence) + gutterSeparator;
+                    context += lineContent + host.getNewLine();
 
                     // Output the gutter and the error span for the line using tildes.
-                    output += formatAndReset(padLeft("", gutterWidth), gutterStyleSequence) + gutterSeparator;
-                    output += redForegroundEscapeSequence;
+                    context += formatAndReset(padLeft("", gutterWidth), gutterStyleSequence) + gutterSeparator;
+                    context += redForegroundEscapeSequence;
                     if (i === firstLine) {
                         // If we're on the last line, then limit it to the last character of the last line.
                         // Otherwise, we'll just squiggle the rest of the line, giving 'slice' no end position.
                         const lastCharForLine = i === lastLine ? lastLineChar : undefined;
 
-                        output += lineContent.slice(0, firstLineChar).replace(/\S/g, " ");
-                        output += lineContent.slice(firstLineChar, lastCharForLine).replace(/./g, "~");
+                        context += lineContent.slice(0, firstLineChar).replace(/\S/g, " ");
+                        context += lineContent.slice(firstLineChar, lastCharForLine).replace(/./g, "~");
                     }
                     else if (i === lastLine) {
-                        output += lineContent.slice(0, lastLineChar).replace(/./g, "~");
+                        context += lineContent.slice(0, lastLineChar).replace(/./g, "~");
                     }
                     else {
                         // Squiggle the entire line.
-                        output += lineContent.replace(/./g, "~");
+                        context += lineContent.replace(/./g, "~");
                     }
-                    output += resetEscapeSequence;
-
-                    output += host.getNewLine();
+                    context += resetEscapeSequence;
                 }
 
                 output += host.getNewLine();
@@ -333,6 +332,12 @@ namespace ts {
             const categoryColor = getCategoryFormat(diagnostic.category);
             const category = DiagnosticCategory[diagnostic.category].toLowerCase();
             output += `${ formatAndReset(category, categoryColor) } TS${ diagnostic.code }: ${ flattenDiagnosticMessageText(diagnostic.messageText, host.getNewLine()) }`;
+
+            if (diagnostic.file) {
+                output += host.getNewLine();
+                output += context;
+            }
+
             output += host.getNewLine();
         }
         return output;

--- a/tests/baselines/reference/prettyContextNotDebugAssertion.errors.txt
+++ b/tests/baselines/reference/prettyContextNotDebugAssertion.errors.txt
@@ -1,8 +1,8 @@
 
+tests/cases/compiler/index.ts(2,1): [91merror[0m TS1005: '}' expected.
+
 [30;47m2[0m 
 [30;47m [0m [91m[0m
-
-tests/cases/compiler/index.ts(2,1): [91merror[0m TS1005: '}' expected.
 
 
 ==== tests/cases/compiler/index.ts (1 errors) ====

